src/Features/CSharp/Portable/CodeFixes/AddExplicitCast/AddExplicitCastCodeFixProvider.cs

```diff
@@ -0,0 +1,372 @@
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeFixes.AddExplicitCast
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp, Name = PredefinedCodeFixProviderNames.AddExplicitCast), Shared]
+    internal sealed partial class AddExplicitCastCodeFixProvider : SyntaxEditorBasedCodeFixProvider
+    {
+        /// <summary>
+        /// CS0266: Cannot implicitly convert from type 'x' to 'y'. An explicit conversion exists (are you missing a cast?)
+        /// </summary>
+        private const string CS0266 = nameof(CS0266);
+
+        /// <summary>
+        /// CS1503: Argument 1: cannot convert from 'x' to 'y'
+        /// </summary>
+        private const string CS1503 = nameof(CS1503);
+        private const int MaximumConversionOptions = 3;
+
+        public sealed override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(CS0266, CS1503);
+
+        internal sealed override CodeFixCategory CodeFixCategory => CodeFixCategory.Compile;
+
+        public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var document = context.Document;
+            var cancellationToken = context.CancellationToken;
+            var diagnostic = context.Diagnostics.First();
+
+            var root = await document.GetRequiredSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var targetNode = root.FindNode(diagnostic.Location.SourceSpan, getInnermostNodeForTie: true)
+                .GetAncestorsOrThis<ExpressionSyntax>().FirstOrDefault();
+            if (targetNode != null)
+            {
+                var hasSolution = GetTypeInfo(semanticModel, root, targetNode, cancellationToken, out var nodeType, out var potentialConversionTypes);
+                if (hasSolution && potentialConversionTypes.Length == 1)
+                {
+                    context.RegisterCodeFix(new MyCodeAction(
+                        CSharpFeaturesResources.Add_explicit_cast,
+                        c => FixAsync(context.Document, context.Diagnostics.First(), c)),
+                        context.Diagnostics);
+                }
+                else if (hasSolution && potentialConversionTypes.Length > 1)
+                {
+                    var actions = ArrayBuilder<CodeAction>.GetInstance();
+
+                    // MaximumConversionOptions: we show at most [MaximumConversionOptions] options for this code fixer
+                    for (var i = 0; i < Math.Min(MaximumConversionOptions, potentialConversionTypes.Length); i++)
+                    {
+                        var convType = potentialConversionTypes[i];
+                        actions.Add(new MyCodeAction(string.Format(CSharpFeaturesResources.Convert_type_to_0, convType.ToDisplayString()),
```

Comment:
instead of adding top level code actions.  We can add a 'inlinable' top level code action `"Convert type"` with all the conversions offered as nested code actions of that.